### PR TITLE
refactor: convert AutoUpdater + CrashReporter to Typescript

### PR DESCRIPTION
**Pull request by MarshallOfSound**
This PR converts the AutoUpdater and CrashReporter APIs to Typescript, please take **a lot** of attention to the AutoUpdater refactor, it's a very sensitive part of the codebase 👍 

Notes: no-notes

### CODE REVIEW

File: lib/browser/api/auto-updater/auto-updater-win.ts [hunk 1]
```diff
@@ -66,8 +67,12 @@ class AutoUpdater extends EventEmitter {
 
   // Private: Emit both error object and message, this is to keep compatibility
   // with Old APIs.
-  emitError (message) {
-    this.emit('error', new Error(message), message)
+  emitError (message: Error | string) {
+    if (message instanceof Error) {
```
Comment (felixrieseberg) [inline lib/browser/api/auto-updater/auto-updater-win.ts]: Nice!

File: lib/browser/api/auto-updater/squirrel-update-win.ts [hunk 2]
```diff
@@ -79,20 +77,25 @@ const spawnUpdate = function (args, detached, callback) {
 }
 
 // Start an instance of the installed app.
-exports.processStart = function () {
+export const processStart = () => {
   return spawnUpdate(['--processStartAndWait', exeName], true, function () {})
 }
 
+type SquirrelUpdateInfo = {
```
Comment (felixrieseberg) [inline lib/browser/api/auto-updater/squirrel-update-win.ts]: We could move that to `Electron.*`, but I think it's fine here, too

Comment (MarshallOfSound) [reply lib/browser/api/auto-updater/squirrel-update-win.ts]: If it was re-used for the macOS impl or something I totally would extract it higher, but this purely lives for this single fn 🤔

**Comment by MarshallOfSound** [comment]
@jkleinsc At the beginning they were intrinsically linked (I can't remember how) but they appear to be slightly less linked now as other PRs doing the same work have landed.  Splitting it out would still result in two PRs that would instantly conflict however so in this case I'd rather we keep it unified.

I can enable the crashReporter specs on all platforms and see how that plays out.  I have manually validated that these files appear to do what they used to do though 😄

**Comment by zcbenz** [comment]
It seems that the crashReporter tests are passing on macOS but not Linux.

@jkleinsc Do you think should we only convert AutoUpdater + CrashReporter files after the tests get fixed in CI?

**Comment by MarshallOfSound** [comment]
So out of date it's not funny, I'll raise a new PR at some point 👍

**Pull request closed without merging**